### PR TITLE
Add ccid apps to the usbip runner

### LINKED ISSUE #140: usbip-runner: Add CCID support

### PR DESCRIPTION
**Pull request by sosthene-nitrokey**
This PR cleans up `usbd-ccid` a bit and adds the CCID apps (opcard) to the usbip runner.

Closes #140, #150 and #151

On my machine I enconter 2 bugs:

- `pcscd` and this somehow manage to trigger a kernel bug:
  1. `cargo run --release --features alpha
  2.  `usbip attach -r localhost -b 1-1` (enough times for it to not return immediately)
  3. `opgpcard status` 
  4. now `lsusb`  hangs and `dmesg` shows an error: `kernel BUG at mm/usercopy.c:101!`

- `gpg` still fails to detect this as an openpgp smartcard for some reason

### CODE REVIEW

**Comment by robin-nitrokey** [comment]
Regarding the first problem:  I think I can reproduce this.  `opgpcard` hangs, and subsequent `usbip` calls don’t seem to work at all.

**Comment by sosthene-nitrokey** [comment]
Does `lsusb` work after that?

**Comment by robin-nitrokey** [comment]
No.

**Comment by sosthene-nitrokey** [comment]
Anything suspicious in `dmesg`?

**Comment by robin-nitrokey** [comment]
Yes, basically the same one you had: 

```
usercopy: Kernel memory exposure attempt detected from SLUB object 'kmalloc-16' (offset 0, size 64)!
kernel BUG at mm/usercopy.c:99
```

**Comment by sosthene-nitrokey** [comment]
There appear to be a similar bug already reported: https://bugzilla.kernel.org/show_bug.cgi?id=215487

**Comment by sosthene-nitrokey** [comment]
I included a fix for #151

**Comment by sosthene-nitrokey** [comment]
Includes a fix for #150

File: components/usbd-ccid/src/pipe.rs [hunk 1]
```diff
@@ -43,6 +43,27 @@ enum Error {
     CommandNotSupported = 0x00,
 }
 
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+enum Outbox {
+    Empty,
+    Full {
+        packet: RawPacket,
+        // Should a zero length packet (zlp) be sent after this packet if the length is maximum
```
Comment (robin-nitrokey) [inline components/usbd-ccid/src/pipe.rs:52]: I think this description is inverted – if the flag is true, we *don’t* want to send a zlp, right?

File: components/usbd-ccid/src/pipe.rs [hunk 2]
```diff
@@ -391,7 +391,8 @@ where
         }
 
         // if let Some(message) = self.interchange.response() {
-        let message: &mut Vec<u8, N> = unsafe { (*self.interchange.interchange.get()).rp_mut() };
+        let message = self.interchange.response().unwrap();
+        // let message: &mut Vec<u8, N> = unsafe { (*self.interchange.interchange.get()).rp_mut() };
```
Comment (robin-nitrokey) [inline components/usbd-ccid/src/pipe.rs:422]: Remove commented code?

**Review by robin-nitrokey — APPROVED** [review]
LGTM!

File: components/apps/src/lib.rs [hunk 3]
```diff
@@ -169,6 +169,13 @@ impl<R: Runner> trussed_usbip::Apps<Client<R>, (&R, NonPortable<R>)> for Apps<R>
     fn with_ctaphid_apps<T>(&mut self, f: impl FnOnce(&mut [&mut dyn CtaphidApp]) -> T) -> T {
         self.ctaphid_dispatch(f)
     }
+
+    fn with_ccid_apps<T>(
+        &mut self,
+        f: impl FnOnce(&mut [&mut dyn apdu_dispatch::App<7609, 7609>]) -> T,
```
Comment (robin-nitrokey) [inline components/apps/src/lib.rs:175]: `ApduApp<ApduCommandSize, ApduResponseSize>`?

File: components/usbd-ccid/src/pipe.rs (hunk 1 above)
Comment (sosthene-nitrokey) [reply components/usbd-ccid/src/pipe.rs]: Indeed. I first added it as a `is_final` but then changed the name and semantics

**Comment by sosthene-nitrokey** [comment]
Ok, so the the way it was done before the fix for #151 is correct, because the response blocks can be sent in multiple parts and it's just that the gnuk test suite doesn't support this.

I still don't understand exactly how my fix made thing worse for the firmware itself, but for now I'll revert it. We should consider fixing the gnuk test suite to use it.

**Comment by sosthene-nitrokey** [comment]
I think we should merge this now. It works well on the firmware itself. I'll then upstream the usbd-ccid changes into https://github.com/trussed-dev/usbd-ccid

**Comment by ZenithalHourlyRate** [comment]
Regarding usbip, I think I have found a bug in your usbip server, see https://lore.kernel.org/lkml/ZBHxfUX60EyCMw5l@Sun/

**Pull request merged**